### PR TITLE
Incremental Hashtag Typehead update

### PR DESCRIPTION
**Pull request by leogrim**
@andrewconner what do you think?

### CODE REVIEW

**Comment by andrewconner** [comment]
Oh sweet, this is awesome. Looking

File: kifi-backend/modules/shoebox/app/com/keepit/typeahead/HashtagTypeahead.scala [hunk 1]
```diff
@@ -14,8 +14,8 @@ import com.keepit.common.db.slick.Database
 import com.keepit.common.time._
 import com.keepit.common.json.TupleFormat
 
-case class UserHashtagTypeahead(ownerId: Id[User], tags: Seq[(Hashtag, Int)], filter: PrefixFilter[Hashtag], createdAt: DateTime) extends PersonalTypeahead[User, Hashtag, (Hashtag, Int)] {
-  def getInfos(tagIds: Seq[Id[Hashtag]]) = Future.successful(tagIds.map(id => tags(id.id.toInt)))
+case class UserHashtagTypeahead(ownerId: Id[User], tags: Seq[(Hashtag, Int)], filter: PrefixFilter[Hashtag], createdAt: DateTime, updatedAt: DateTime) extends PersonalTypeahead[User, Hashtag, (Hashtag, Int)] {
+  def getInfos(tagIds: Seq[Id[Hashtag]]) = Future.successful(tagIds.map(id => tags(id.id.toInt)).filter(_._2 > 0))
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/typeahead/HashtagTypeahead.scala:18]: :+1: nice catch

File: kifi-backend/modules/shoebox/app/com/keepit/typeahead/HashtagTypeahead.scala [hunk 2]
```diff
@@ -59,12 +76,31 @@ class HashtagTypeahead @Inject() (
...
+        }
+        val newTags = (updatesByNormalizedTag -- typeahead.tags.map(_._1.normalized)).values
+        // Actually reordering tags and rebuilding the filter from scratch. If we don't reorder we can append, worth it?
+        val (sortedTags, filter) = sortTagsAndBuildFilter(userId, updatedTags ++ newTags)
+        val updatedTypeahead = UserHashtagTypeahead(userId, sortedTags, filter, typeahead.createdAt, clock.now())
+        invalidate(updatedTypeahead)
+    }
+  }
 
+  protected def extractName(hashtagWithKeepCount: (Hashtag, Int)) = hashtagWithKeepCount._1.tag
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/typeahead/HashtagTypeahead.scala:99]: Should this be normalized, or is it already that way?

**Comment by andrewconner** [comment]
Looks good to me. Will this ever need defragmenting?

File: kifi-backend/modules/shoebox/app/com/keepit/typeahead/HashtagTypeahead.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/typeahead/HashtagTypeahead.scala:99]: I don't think it matters as it's gonna be tokenized

**Comment by leogrim** [comment]
Yeah I guess if we find out it's not so reliable, we can have a scheduled defragmentation

**Pull request merged**